### ISSUE TITLE
How to handle overflow issue

### ISSUE THREAD

**Issue by RoiEXLab** [triplea-game/triplea#2376]
I discovered an issue with @prastle earlier that day, which seems like a Java bug to me (probably worth reporting, if I knew where).
An ArithmeticException is being thrown when banning someone forever using the MenuBar Toolbox.
The change in https://github.com/triplea-game/triplea/commit/62ede44a596452aca66e9895545c6268a5d3375a#diff-8720367863d7314b908593d8073c7baeR158 introduced it, but did not really cause it.
Although both snippets technically should do the same, they do not, because long is a primitive type.
`requestTimespanSupplication()` returns `Long.MAX_VALUE` if the mod/admin bans somebody forever. when adding this value to `System.currentTimeMillis()`, the long silently overflows resulting in a negative value, which was no problem, because `new Date(long)` does treat longs as unsigned.
The Problem is, that Instant does not.
Because `Instant.toEpochMillis()` throws an ArithmeticException when overflowing we get the observed error, although Date could handle this negative value, e.g. Date could represent the given Instant to `Date.from`.
Therefore `System.out.println(Date.from(Instant.now().plusMillis(Long.MAX_VALUE)));` throws an exception instead of printing a valid date.

The question is how we should tackle this issue. We should be able to workaround this issue by copying the Math from the `Instant.toEpochMilli` method:
```
public long toEpochMilli() {
        if (seconds < 0 && nanos > 0) {
            long millis = Math.multiplyExact(seconds+1, 1000);
            long adjustment = nanos / 1000_000 - 1000;
            return Math.addExact(millis, adjustment);
        } else {
            long millis = Math.multiplyExact(seconds, 1000);
            return Math.addExact(millis, nanos / 1000_000);
        }
    }
```
,replacing Math.multiplyExact with '*' and addExact with '+', so we get
```
public long toEpochMilli() {
        if (seconds < 0 && nanos > 0) {
            long millis = (seconds+1) * 1000;
            long adjustment = nanos / 1000_000 - 1000;
            return millis + adjustment;
        } else {
            long millis = seconds * 1000;
            return millis + (nanos / 1000_000);
        }
    }
```
*BUT* this solutions actually doesn't makes use of the _real_ forever banning. In this case forever is just a point a couple thousand years in the future, not NULL like in the other cases.

TL;DR We basically have 2 options:
1. Workaround this issue
2. Rework class to be consistent with other classes

**Comment by ssoloff** [comment]
I only looked at this briefly, so I may be misunderstanding the issue, but it appears `requestTimespanSupplication()` is returning three different types of values:

1. a relative duration to be added to an absolute time (values in the range [0, `Long.MAX_VALUE`)),
1. an absolute time representing "impossibly far in the future" (the value `Long.MAX_VALUE`), and
1. a flag indicating the user canceled the ban/mute operation (the value -1).

The easiest fix would be to always return a consistent type of value.  To avoid the overflow issue described above, it would seem that always returning an absolute time would work best.  To accommodate case (3), that value could be wrapped in an `Optional`.  I've attached a [patch](https://github.com/triplea-game/triplea/files/1312277/issue-2376.patch.txt) (untested) that attempts to do this.  Below is the patch with whitespace changes ignored:

```diff
diff --git a/src/main/java/games/strategy/triplea/ui/menubar/LobbyMenu.java b/src/main/java/games/strategy/triplea/ui/menubar/LobbyMenu.java
index 5d47720..8d35d29 100644
--- a/src/main/java/games/strategy/triplea/ui/menubar/LobbyMenu.java
+++ b/src/main/java/games/strategy/triplea/ui/menubar/LobbyMenu.java
@@ -9,6 +9,7 @@ import java.time.Instant;
 import java.util.ArrayList;
 import java.util.Date;
 import java.util.List;
+import java.util.Optional;
 
 import javax.swing.JButton;
 import javax.swing.JCheckBoxMenuItem;
@@ -156,7 +157,7 @@ public class LobbyMenu extends JMenuBar {
           return;
         }
       }
-      final Instant expire = Instant.now().plusMillis(requestTimespanSupplication());
+      requestTimespanSupplication(Instant.now()).ifPresent(expire -> {
         final IModeratorController controller = (IModeratorController) lobbyFrame.getLobbyClient().getMessengers()
             .getRemoteMessenger().getRemote(ModeratorController.getModeratorControllerName());
         try {
@@ -165,6 +166,7 @@ public class LobbyMenu extends JMenuBar {
           ClientLogger.logQuietly(ex);
         }
       });
+    });
     item.setEnabled(true);
     parentMenu.add(item);
   }
@@ -197,7 +199,7 @@ public class LobbyMenu extends JMenuBar {
           return;
         }
       }
-      final Instant expire = Instant.now().plusMillis(requestTimespanSupplication());
+      requestTimespanSupplication(Instant.now()).ifPresent(expire -> {
         final IModeratorController controller = (IModeratorController) lobbyFrame.getLobbyClient().getMessengers()
             .getRemoteMessenger().getRemote(ModeratorController.getModeratorControllerName());
         try {
@@ -207,6 +209,7 @@ public class LobbyMenu extends JMenuBar {
           ClientLogger.logQuietly(ex);
         }
       });
+    });
     item.setEnabled(true);
     parentMenu.add(item);
   }
@@ -280,7 +283,7 @@ public class LobbyMenu extends JMenuBar {
     parentMenu.add(item);
   }
 
-  private long requestTimespanSupplication() {
+  private Optional<Instant> requestTimespanSupplication(final Instant now) {
     final List<String> timeUnits = new ArrayList<>();
     timeUnits.add("Minute");
     timeUnits.add("Hour");
@@ -294,22 +297,22 @@ public class LobbyMenu extends JMenuBar {
             JOptionPane.OK_CANCEL_OPTION, JOptionPane.QUESTION_MESSAGE, null, timeUnits.toArray(),
             timeUnits.toArray()[3]);
     if (result < 0) {
-      return -1;
+      return Optional.empty();
     }
     final String selectedTimeUnit = (String) timeUnits.toArray()[result];
     if (selectedTimeUnit.equals("Forever")) {
-      return Long.MAX_VALUE;
+      return Optional.of(Instant.MAX);
     }
     final String stringr = JOptionPane.showInputDialog(
         lobbyFrame,
         "Now please enter the length of time: (In " + selectedTimeUnit + "s) ",
         1);
     if (stringr == null) {
-      return -1;
+      return Optional.empty();
     }
     final long result2 = Long.parseLong(stringr);
     if (result2 < 0) {
-      return -1;
+      return Optional.empty();
     }
     long ticks = 0;
     if (selectedTimeUnit.equals("Minute")) {
@@ -325,7 +328,7 @@ public class LobbyMenu extends JMenuBar {
     } else if (selectedTimeUnit.equals("Year")) {
       ticks = result2 * 1000 * 60 * 60 * 24 * 365;
     }
-    return ticks;
+    return Optional.of(now.plusMillis(ticks));
   }
 
   private void createSettingsMenu(final LobbyMenu menuBar) {
```

Note that I'm not sure if case (3) was ever handled properly.  It seems that no change should be made to the database when the user cancels the dialog or enters an invalid value; previously the -1 would simply be added to the current time in milliseconds and sent along as the ban/mute expiration time to `UserController`.  The above patch does not update the database if `requestTimespanSupplication()` returns an empty `Optional` (which corresponds to -1 being returned in the previous implementation).

**Comment by RoiEXLab** [comment]
@ssoloff I'd like to encourage you open a PR for this issue, I just did some testing with Date and instance objects.
You found a pretty good solution for the problem, but I suggest passing null to the banUsername/banMac methods if the ban is forever in order to be consistent.

**Comment by ssoloff** [comment]
Well, that's probably not going to happen any time soon. :smile:  I have too many things in-flight at the moment, and this would require some significant testing and test writing.

> I suggest passing null to the banUsername/banMac methods if the ban is forever in order to be consistent

Returning `null` from `requestTimespanSupplication()` would cause some confusion between it and an empty `Optional`.  Maybe the thing to do is use `Instant#MAX` as a sentinel value and transform it to `null` in `LobbyMenu` before calling banXXX?  IDK, that still smells a bit.

Maybe the database schema is where the smell is coming from, and we should avoid using `null` to represent _forever_ and instead use `Instant#MAX`?  The Javadocs for `Instant#MAX` even say, "This could be used by an application as a 'far future' instant."  It seems that if we avoided using `null`, it would simplify the logic in the ban/mute controllers significantly without changing any behavior.  Did you happen to consider this as part of #2355?

**Comment by RoiEXLab** [comment]
@ssoloff I did not, but my concern would be that Instant.MAX is actually too big of a value that postgres could handle.
java.sql.Timestamp extends java.util.Date, so we wouldn't be able to convert types already at that point.

**Comment by RoiEXLab** [comment]
Anyways I do not want to push the scope of #2355 even further. This would be a change for a follow-up PR, where we could also consider merging banned_macs and muted_macs tables into punished_macs with 3 rows as well as merging banned_usernames and muted_usernames together into one table as well.

**Comment by ssoloff** [comment]
> my concern would be that Instant.MAX is actually too big of a value that postgres could handle

Good point.  And it looks like these methods will throw an `IllegalArgumentException` if an `Instant` value larger than the maximum `Timestamp` value is passed.  Since we changed these methods from previously accepting a `Date` to now accepting an `Instant`, do we need to do some additional mapping between the domain and data layers.  For example, map any `Instant` value greater than the maximum `Timestamp` value to the maximum `Timestamp` value?  Or is throwing an exception the appropriate behavior?

If the latter, then an alternative could be to just document in the controller methods that `Instant#MAX` represents _forever_, and map that single `Instant` value to a `null` `Timestamp` value (and vice versa) in the controller implementation:

```java
ps.setTimestamp(2, Instant.MAX.equals(banTill) ? null : Timestamp.from(banTill);

// . . .

final Timestamp timestamp = rs.getTimestamp(2);
final Instant banTill = timestamp != null ? timestamp.toInstant() : Instant.MAX;
```

So even though the database column is still nullable, the logic would always deal with a non-`null` `Instant`.

**Comment by RoiEXLab** [comment]
Something like this would be desirable, but then again this would be inconsistent with null being passed into the ban controller method. 
Changing that would introduce breaking changes.

**Comment by RoiEXLab** [comment]
I created a bug report btw: http://bugs.java.com/bugdatabase/view_bug.do?bug_id=JDK-8187708
